Feature/edit search term

Komock opened the pull request:
#1489

Komock commented:
@varnastadeus Hi, could you check this?

hade94 commented:
I am waiting for this feature too :)

varnastadeus commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -110,6 +110,7 @@ export class NgSelectComponent implements OnDestroy, OnChanges, AfterViewInit, C
     @Input() readonly = false;
     @Input() searchWhileComposing = true;
     @Input() minTermLength = 0;
+    @Input() editSearchTerm = false;
```
maybe rename to `editableSearchTerm`?

varnastadeus commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -570,6 +580,10 @@ export class NgSelectComponent implements OnDestroy, OnChanges, AfterViewInit, C
         }
 
+        if (this._editSearchTerm) {
+            this.searchTerm = (this.selectedItems && this.selectedItems[0] && this.selectedItems[0].label) || null;
```
```suggestion
           const selected = this.selectedItems && this.selectedItems[0];
           this.searchTerm = selected ? selected.label : null;
```

Komock commented:
> Thanks for your contribution 🎉I added some comments. Also, there is additional bug which should be fixed before merge:
> 
> 1. Select item -> "Karyn Wright"
> 2. Clear some chars -> "Karyn"
> 3. Close the dropdown
>    I would expect "Karyn Wright" should remain since it is selected item - currently "Karyn" shows as selected and after focus changes back to original.

@varnastadeus Hi. I fixed this bug, now I also update `searchTerm` on blur and `toggleItem()`. And `@Input` property also renamed to `editableSearchTerm`.

varnastadeus commented:
> > Thanks for your contribution 🎉I added some comments. Also, there is additional bug which should be fixed before merge:
> > 
> > 1. Select item -> "Karyn Wright"
> > 2. Clear some chars -> "Karyn"
> > 3. Close the dropdown
> >    I would expect "Karyn Wright" should remain since it is selected item - currently "Karyn" shows as selected and after focus changes back to original.
> 
> @varnastadeus Hi. I fixed this bug, now I also update `searchTerm` on blur and `toggleItem()`. And `@Input` property also renamed to `editableSearchTerm`.

When deleting whole search term, selected item reappears. Maybe this shouldn't happen in case of `editableSearchTerm`?

Komock commented:
> > > Thanks for your contribution 🎉I added some comments. Also, there is additional bug which should be fixed before merge:
> > > 
> > > 1. Select item -> "Karyn Wright"
> > > 2. Clear some chars -> "Karyn"
> > > 3. Close the dropdown
> > >    I would expect "Karyn Wright" should remain since it is selected item - currently "Karyn" shows as selected and after focus changes back to original.
> > 
> > 
> > @varnastadeus Hi. I fixed this bug, now I also update `searchTerm` on blur and `toggleItem()`. And `@Input` property also renamed to `editableSearchTerm`.
> 
> When deleting whole search term, selected item reappears. Maybe this shouldn't happen in case of `editableSearchTerm`?

i agree, it looks buggy. I added `clearModel()` if `searchTerm` empty

hade94 commented:
@varnastadeus Could you check this please?

varnastadeus commented:
@Komock can you revert the last commit? I have checked default behavior when clearing search term selected item also reappears so let's keep it same here. Sorry for inconvenience

Komock commented:
> @Komock can you revert the last commit? I have checked default behavior when clearing search term selected item also reappears so let's keep it same here. Sorry for inconvenience

It's ready

varnastadeus commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -439,16 +445,24 @@ export class NgSelectComponent implements OnDestroy, OnChanges, AfterViewInit, C
                 this._clearSearch();
             }
 
+            if (this._editableSearchTerm) {
```
does this really needed? because you are doing the same in method above `toggleItem` with:

```js
if (this._editableSearchTerm) {
            this._setSearchTermFromItems();
}
```

varnastadeus commented:
@Komock I added 1 more comment

Komock commented:
> @Komock I added 1 more comment

Fixed.
And I have fixed one more with `resetFilteredItems`. 
- select item
- type wrong search term
- blur
- focus again
Dropdown list empty

varnastadeus commented:
> > @Komock I added 1 more comment
> 
> Fixed.
> And I have fixed one more with `resetFilteredItems`.
> 
> * select item
> * type wrong search term
> * blur
> * focus again
>   Dropdown list empty

Please add that case as a unit test or update your test case so we won't get in trouble in the future 👍

Komock commented:
> > > @Komock I added 1 more comment
> > 
> > 
> > Fixed.
> > And I have fixed one more with `resetFilteredItems`.
> > 
> > * select item
> > * type wrong search term
> > * blur
> > * focus again
> >   Dropdown list empty
> 
> Please add that case as a unit test or update your test case so we won't get in trouble in the future 👍

Done!
